connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java

```diff
@@ -70,60 +73,105 @@
-            throw new ConfigException(SinkTask.TOPICS_CONFIG + " and " + SinkTask.TOPICS_REGEX_CONFIG +
-                " are mutually exclusive options, but both are set.");
+            String errorMessage = TOPICS_CONFIG + " and " + TOPICS_REGEX_CONFIG + " are mutually exclusive options, but both are set.";
+            onError.accept(new ConfigError(TOPICS_CONFIG, topicsList, errorMessage));
+            onError.accept(new ConfigError(TOPICS_REGEX_CONFIG, topicsRegex, errorMessage));
         }
 
         if (!hasTopicsConfig && !hasTopicsRegexConfig) {
-            throw new ConfigException("Must configure one of " +
-                SinkTask.TOPICS_CONFIG + " or " + SinkTask.TOPICS_REGEX_CONFIG);
+            String errorMessage = "Must configure one of " + TOPICS_CONFIG + " or " + TOPICS_REGEX_CONFIG;
+            onError.accept(new ConfigError(TOPICS_CONFIG, topicsList, errorMessage));
+            onError.accept(new ConfigError(TOPICS_REGEX_CONFIG, topicsRegex, errorMessage));
         }
 
         if (hasDlqTopicConfig) {
-            String dlqTopic = props.get(DLQ_TOPIC_NAME_CONFIG).trim();
             if (hasTopicsConfig) {
                 List<String> topics = parseTopicsList(props);
                 if (topics.contains(dlqTopic)) {
-                    throw new ConfigException(String.format("The DLQ topic '%s' may not be included in the list of "
-                            + "topics ('%s=%s') consumed by the connector", dlqTopic, SinkTask.TOPICS_REGEX_CONFIG, topics));
+                    String errorMessage = String.format(
+                        "The DLQ topic '%s' may not be included in the list of topics ('%s=%s') consumed by the connector",
+                        dlqTopic, TOPICS_CONFIG, topics
+                    );
+                    onError.accept(new ConfigError(TOPICS_CONFIG, topicsList, errorMessage));
                 }
             }
             if (hasTopicsRegexConfig) {
-                String topicsRegexStr = props.get(SinkTask.TOPICS_REGEX_CONFIG);
-                Pattern pattern = Pattern.compile(topicsRegexStr);
+                Pattern pattern = Pattern.compile(topicsRegex);
                 if (pattern.matcher(dlqTopic).matches()) {
-                    throw new ConfigException(String.format("The DLQ topic '%s' may not be included in the regex matching the "
-                            + "topics ('%s=%s') consumed by the connector", dlqTopic, SinkTask.TOPICS_REGEX_CONFIG, topicsRegexStr));
+                    String errorMessage = String.format(
+                        "The DLQ topic '%s' may not be included in the regex matching the topics ('%s=%s') consumed by the connector",
+                        dlqTopic, TOPICS_REGEX_CONFIG, topicsRegex
+                    );
+                    onError.accept(new ConfigError(TOPICS_REGEX_CONFIG, topicsRegex, errorMessage));
                 }
             }
         }
     }
 
+    private static class ConfigError {
```

Comment:
> I think the reason you're using this class instead of the plain exception to avoid the cost of instantiating the exception and filling the stacktrace.

Not exactly; it's because the `ConfigException` class isn't really an effective POJO for a property, its value, and an error message. There are no getters for any of those fields, and two of the three constructors available would permit null values for some of them. I chose not to change the `ConfigException` class to add these getters since I'm not sure it's the best idea for an exception class to be doing double duty as a POJO as well.

> I think passing in Map<String, ConfigValue> and calling the addErrorMessage function as a utility function directly avoids the need for this POJO, while getting you the ability to return multiple errors.

This is true, but I'm not sure how we could make that work with the other current use for `SinkConnectorConfig::validate`, which is to do a [preliminary validation step](https://github.com/apache/kafka/blob/db42afd6e24ef4291390b4d1c1f10758beedefed/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L295) during `WorkerSinkTask` initialization.

> Also, Is there a reason that we can't have this validation return a ConfigInfos object, and merge it in late like the rest of the validations taking place in AbstractHerder::validateConnnectorConfig? That seems like the standard convention.

It'd make it a little more difficult for subclasses (i.e., `DistributedHerder`) to override behavior and add their own logic on top.